Keep databse consistency between source item and low stocknotificatio…

phoenix128 opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
Database consistency of low level stock notification table is now moved at application level.
This implementation covers the bulk operations.

### Fixed Issues (if relevant)
1. magento-engcom/msi#1659: Something Went Wrong error during the SourceItem migration 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

phoenix128 committed:
Keep databse consistency between source item and low stocknotification item via plugin

phoenix128 commented:
Seems like tests are failing for an external reason.

phoenix128 committed:
Reworked plugin replacing "around" with "after"

maghamed committed:
MSI-1659: fix static test

maghamed commented on app/code/Magento/InventoryLowQuantityNotification/Plugin/InventoryApi/SourceItemsDeleteInterfacePlugin.php:
```diff
@@ -0,0 +1,51 @@
+        foreach ($sourceItems as $sourceItem) {
+            /** @var SourceItemInterface $sourceItem */
+            $this->delete->execute($sourceItem->getSourceCode(), $sourceItem->getSku());
+        }
```
a little bit worrying about potential performance degradation. 
As we are pluginizaing Bulk operation with logic which process records one by one, 
this is current issue in LowStock Design, but here should be another bulk operation being called

phoenix128 replied:
Hello @maghamed ,
actually this plugin is for `Magento\InventoryApi\Api\SourceItemsDeleteInterface` that is not involved in bulk operations. 

The plugin handling the same operation at bulk level is: `Magento\InventoryLowQuantityNotification\Plugin\InventoryCatalogApi\BulkSourceUnassignInterfacePlugin`. As you can see this last one operates at resource level.

maghamed replied:
Just noticed that I did not answer in this thread and we still have this PR opened.

Looks like we have always some confusion regarding the terminology, when I say `Bulk` API I mean the API which accepts a bunch of entities (an array), but not a single entity like most of the Service Contracts we have (i.e. all Repositories). But looks like you consider `Bulk API` term as an async execution feature.
 
For example, 
ProductRepository::save is a non-bulk API , as it can handle just one product entity per a call.
```
interface ProductRepositoryInterface
{
    /**
     * Create product
     *
     * @param \Magento\Catalog\Api\Data\ProductInterface $product
     * @param bool $saveOptions
     * @return \Magento\Catalog\Api\Data\ProductInterface
     * @throws \Magento\Framework\Exception\InputException
     * @throws \Magento\Framework\Exception\StateException
     * @throws \Magento\Framework\Exception\CouldNotSaveException
     */
    public function save(\Magento\Catalog\Api\Data\ProductInterface $product, $saveOptions = false);
```
SourceItemsDeleteInterface - is a bulk API, as it accepts an array of sourceItems
```
interface SourceItemsDeleteInterface
{
    /**
     * Delete Multiple Source item data
     *
     * @param \Magento\InventoryApi\Api\Data\SourceItemInterface[] $sourceItems
     * @return void
     * @throws \Magento\Framework\Exception\InputException
     * @throws \Magento\Framework\Exception\CouldNotSaveException
     */
    public function execute(array $sourceItems): void;
}

```

So, in this particular situation we have a service 
`SourceItemsDeleteInterface ::execute(array $sourceItems)`
which accepts an array of entities to work with. 

But in the plugin which handles this logic we have an iteration 

```
        foreach ($sourceItems as $sourceItem) {
            /** @var SourceItemInterface $sourceItem */
            $this->delete->execute($sourceItem->getSourceCode(), $sourceItem->getSku());
        }
```
calling ONE api call for each single entity from provided input array. 
which could be quite time consuming if the array is big enough.


Moreover you are using private implementation
`Magento\InventoryLowQuantityNotification\Model\ResourceModel\SourceItemConfiguration\Delete`
, but not the public API from InventoryLowQuantityNotificationApi module
`\Magento\InventoryLowQuantityNotificationApi\Api\DeleteSourceItemConfigurationInterface`
```
namespace Magento\InventoryLowQuantityNotificationApi\Api;

/**
 * Delete the source item configuration
 *
 * @api
 */
interface DeleteSourceItemConfigurationInterface
{
    /**
     * @param string $sourceCode
     * @param string $sku
     * @return void
     */
    public function execute(string $sourceCode, string $sku): void;
}

```

The interesting thing is that there is Bulk Service in the LowStockQuantity module 
`\Magento\InventoryLowQuantityNotification\Model\ResourceModel\BulkConfigurationUnassign::execute`
but it's not public API of the module.

phoenix128 replied:
Ok, clear now

phoenix128 replied:
Ok, clear now. I'm reworking it.

phoenix128 committed:
Rework DeleteSourceItemConfigurationInterface to accept multiple source items
\Magento\InventoryLowQuantityNotification\Model\ResourceModel\SourceItemConfiguration\DeleteMultiple is reworked to accept multiple source items per time.
A single SQL query is performed to guarantee performance. Test have been adapted and all the methods calling the old implementation have been adapted.
API call was moved from DELETE to POST on a different URL to allow a multiple source items payload

phoenix128 committed:
Merge branch '2.3-develop' into issue-1659-database-soft-consistency-for-low-stock-notification

phoenix128 committed:
Merge of feature branch with DeleteMultiple approach

phoenix128 committed:
Missing return type

maghamed commented on app/code/Magento/InventoryLowQuantityNotificationApi/etc/webapi.xml:
```diff
@@ -18,8 +18,8 @@
-    <route url="/V1/inventory/low-quantity-notification/:sourceCode/:sku" method="DELETE">
-        <service class="Magento\InventoryLowQuantityNotificationApi\Api\DeleteSourceItemConfigurationInterface" method="execute"/>
+    <route url="/V1/inventory/low-quantity-notifications-delete" method="POST">
+        <service class="Magento\InventoryLowQuantityNotificationApi\Api\DeleteSourceItemsConfigurationInterface" method="execute"/>
```
please fix corresponding web API test
```
namespace Magento\InventoryLowQuantityNotificationApi\Test\Api;

use Magento\InventoryLowQuantityNotificationApi\Api\Data\SourceItemConfigurationInterface;
use Magento\TestFramework\TestCase\WebapiAbstract;
use Magento\Framework\Webapi\Rest\Request;

class DeleteSourceItemConfigurationTest extends WebapiAbstract
{
    const RESOURCE_PATH = '/V1/inventory/low-quantity-notification';
    const SERVICE_NAME_DELETE = 'inventoryLowQuantityNotificationApiDeleteSourceItemConfigurationV1';
    const SERVICE_NAME_GET = 'inventoryLowQuantityNotificationApiGetSourceItemConfigurationV1';

    /**
     * @magentoApiDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
     * @magentoApiDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
     * @magentoApiDataFixture ../../../../app/code/Magento/InventoryLowQuantityNotificationApi/Test/_files/source_item_configuration.php
     */
    public function testDeleteSourceItemConfiguration()
    {
        $sourceCode = 'eu-1';
        $sku = 'SKU-1';

        $serviceInfo = [
            'rest' => [
                'resourcePath' => self::RESOURCE_PATH . '/' . $sourceCode . '/' . $sku,
                'httpMethod' => Request::HTTP_METHOD_DELETE,
            ],
            'soap' => [
                'service' => self::SERVICE_NAME_DELETE,
                'operation' => self::SERVICE_NAME_DELETE . 'Execute',
            ],
        ];
```

maghamed commented:
Run Jenkins build 2.3.0-release
https://m2build.devops.magento.com/job/Existing-PR-Tests/8573/

2.3-develop Magento CE branch
https://m2build.devops.magento.com/job/Existing-PR-Tests/8574/

phoenix128 committed:
FIX failing test due to rework of MassDelete resource model
